File: reference/python/pcam_runtime/networking.py
"""Bounded lockstep coordination and server-authoritative correction planning."""

from __future__ import annotations

from dataclasses import dataclass

from .canonical import canonical_dumps, canonical_hash
from .immutable import freeze_value


@dataclass(frozen=True)
class LockstepTick:
    status: str
    tick: int
    tick_document: dict[str, object] | None = None
    missing_peers: tuple[str, ...] = ()
    predicted_peers: tuple[str, ...] = ()
    digest_due: bool = False


@dataclass(frozen=True)
class DigestResolution:
    status: str
    tick: int
    mismatched_peers: tuple[str, ...] = ()


@dataclass(frozen=True)
class ServerCorrectionPlan:
    operation: str
    authoritative_tick: int
    discard_prediction_ticks: int


@dataclass(frozen=True)
class _PeerPacket:
    definition_set_hash: str
    inputs: tuple[dict[str, object], ...]
    contacts: tuple[dict[str, object], ...]

    def __post_init__(self) -> None:
        object.__setattr__(self, "inputs", freeze_value(self.inputs))
        object.__setattr__(self, "contacts", freeze_value(self.contacts))


class LockstepCoordinator:
    def __init__(
        self,
        required_peers: tuple[str, ...],
        definition_set_hash: str,
        input_availability_policy: str,
        digest_interval_ticks: int,
        desynchronization_policy: str,
        predictor_id: str | None = None,
    ):
        canonical_peers = tuple(sorted(required_peers, key=lambda item: item.encode("utf-8")))
        if not canonical_peers or len(canonical_peers) != len(set(canonical_peers)):
            raise ValueError("lockstep peers must be nonempty and unique")
        if input_availability_policy not in {"WAIT", "PREDICT"}:
            raise ValueError("unsupported lockstep input availability policy")
        if input_availability_policy == "PREDICT" and predictor_id != "pcam.predict.no-input.v1":
            raise ValueError("bounded lockstep prediction requires pcam.predict.no-input.v1")
        if digest_interval_ticks <= 0:
            raise ValueError("digest interval must be positive")
        if desynchronization_policy not in {"pcam.desync.abort", "pcam.desync.report"}:
            raise ValueError("unsupported desynchronization policy")
        self.required_peers = canonical_peers
        self.definition_set_hash = definition_set_hash
        self.input_availability_policy = input_availability_policy
        self.digest_interval_ticks = digest_interval_ticks
        self.desynchronization_policy = desynchronization_policy
        self.predictor_id = predictor_id
        self.next_tick = 0
        self.aborted = False
        self._packets: dict[str, _PeerPacket] = {}
        self._digest_ticks: set[int] = set()
        self._digest_reports: dict[int, dict[str, str]] = {}
        self._local_digests: dict[int, str] = {}

    def submit(
        self,
        peer_id: str,
        tick: int,
        definition_set_hash: str,
        inputs: tuple[dict[str, object], ...] = (),
        contacts: tuple[dict[str, object], ...] = (),
    ) -> None:
        self._require_active()
        if peer_id not in self.required_peers:
            raise ValueError("unknown lockstep peer")
        if tick != self.next_tick:
            raise ValueError("lockstep packet tick does not match next tick")
        if definition_set_hash != self.definition_set_hash:
            raise ValueError("lockstep definition-set hash mismatch")
        if peer_id in self._packets:
            raise ValueError("duplicate lockstep peer packet")
        if any(int(item.get("assigned_tick", -1)) != tick for item in inputs):
            raise ValueError("lockstep input assigned_tick mismatch")
        self._packets[peer_id] = _PeerPacket(definition_set_hash, inputs, contacts)

    def advance(self) -> LockstepTick:
        self._require_active()
        missing = tuple(peer for peer in self.required_peers if peer not in self._packets)
        if missing and self.input_availability_policy == "WAIT":
            return LockstepTick("WAITING", self.next_tick, missing_peers=missing)
        packets = tuple(self._packets[peer] for peer in self.required_peers if peer in self._packets)
        host_hashes = {canonical_hash(packet.contacts) for packet in packets}
        if len(host_hashes) > 1:
            raise ValueError("lockstep deterministic host snapshot mismatch")
        contacts = packets[0].contacts if packets else ()
        inputs = tuple(item for packet in packets for item in packet.inputs)
        identifiers = [str(item.get("input_id")) for item in inputs]
        if len(identifiers) != len(set(identifiers)):
            raise ValueError("lockstep input_id collision")
        inputs = tuple(sorted(inputs, key=_input_key))
        tick = self.next_tick
        due = (tick + 1) % self.digest_interval_ticks == 0
        document = {"inputs": list(inputs), "contacts": list(contacts)}
        self._packets.clear()
        self.next_tick += 1
        if due:
            self._digest_ticks.add(tick)
        return LockstepTick(
            "READY",
            tick,
            freeze_value(document),
            predicted_peers=missing,
            digest_due=due,
        )

    def submit_digest(
        self,
        peer_id: str,
        tick: int,
        reported_digest: str,
        local_digest: str,
    ) -> DigestResolution:
        self._require_active()
        if peer_id not in self.required_peers:
            raise ValueError("unknown lockstep peer")
        if tick not in self._digest_ticks:
            raise ValueError("state digest is not due for tick")
        known_local = self._local_digests.setdefault(tick, local_digest)
        if known_local != local_digest:
            raise ValueError("local digest changed during exchange")
        reports = self._digest_reports.setdefault(tick, {})
        if peer_id in reports:
            raise ValueError("duplicate peer digest")
        reports[peer_id] = reported_digest
        missing = set(self.required_peers) - set(reports)
        if missing:
            return DigestResolution("WAITING", tick)
        mismatched = tuple(peer for peer in self.required_peers if reports[peer] != local_digest)
        if mismatched and self.desynchronization_policy == "pcam.desync.abort":
            self.aborted = True
            return DigestResolution("ABORTED", tick, mismatched)
        return DigestResolution("REPORTED" if mismatched else "MATCH", tick, mismatched)

    def _require_active(self) -> None:
        if self.aborted:
            raise ValueError("lockstep session is aborted")


class ServerAuthoritativeCorrectionPlanner:
    def __init__(self, correction_policy: str, max_latency_compensation_ticks: int):
        if correction_policy not in {
            "pcam.correction.resimulate.v1",
            "pcam.correction.replace-discard.v1",
        }:
            raise ValueError("unsupported server correction policy")
        if max_latency_compensation_ticks < 0:
            raise ValueError("correction limit must be nonnegative")
        self.correction_policy = correction_policy
        self.max_latency_compensation_ticks = max_latency_compensation_ticks

    def plan(
        self,
        current_tick: int,
        authoritative_tick: int,
        complete_state_supplied: bool,
    ) -> ServerCorrectionPlan:
        if authoritative_tick < 0 or authoritative_tick > current_tick:
            raise ValueError("authoritative correction tick is invalid")
        discard = current_tick - authoritative_tick
        if discard > self.max_latency_compensation_ticks:
            raise ValueError("authoritative correction exceeds compensation limit")
        if self.correction_policy == "pcam.correction.replace-discard.v1":
            if not complete_state_supplied:
                raise ValueError("replace-discard correction requires complete authoritative state")
            operation = "REPLACE_AND_DISCARD"
        else:
            operation = "RESTORE_AND_RESIMULATE"
        return ServerCorrectionPlan(operation, authoritative_tick, discard)


def _input_key(value: dict[str, object]) -> tuple[object, ...]:
    return (
        int(value.get("source_entity_id", 0)),
        int(value.get("sequence", 0)),
        str(value.get("command_id", "")).encode("utf-8"),
        str(value.get("input_id", "")).encode("utf-8"),
        canonical_dumps(value),
    )

File: independent/rust/src/networking.rs
use crate::{CanonicalError, canonical_hash, canonicalize};
use serde_json::{Value, json};
use std::collections::{BTreeMap, BTreeSet};

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum NetworkError {
    InvalidProfile,
    Aborted,
    UnknownPeer,
    TickMismatch,
    DefinitionMismatch,
    DuplicatePacket,
    InputTickMismatch,
    HostMismatch,
    DuplicateInput,
    DigestNotDue,
    DuplicateDigest,
    LocalDigestChanged,
    InvalidCorrection,
    CorrectionWindowExceeded,
    CompleteStateRequired,
    Canonical(CanonicalError),
}

impl From<CanonicalError> for NetworkError {
    fn from(error: CanonicalError) -> Self {
        Self::Canonical(error)
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct LockstepTick {
    pub status: String,
    pub tick: u64,
    pub tick_document: Option<Value>,
    pub missing_peers: Vec<String>,
    pub predicted_peers: Vec<String>,
    pub digest_due: bool,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct DigestResolution {
    pub status: String,
    pub tick: u64,
    pub mismatched_peers: Vec<String>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ServerCorrectionPlan {
    pub operation: String,
    pub authoritative_tick: u64,
    pub discard_prediction_ticks: u64,
}

#[derive(Debug, Clone)]
struct PeerPacket {
    inputs: Vec<Value>,
    contacts: Vec<Value>,
}

#[derive(Debug, Clone)]
pub struct LockstepCoordinator {
    required_peers: Vec<String>,
    definition_set_hash: String,
    input_availability_policy: String,
    digest_interval_ticks: u64,
    desynchronization_policy: String,
    pub next_tick: u64,
    pub aborted: bool,
    packets: BTreeMap<String, PeerPacket>,
    digest_ticks: BTreeSet<u64>,
    digest_reports: BTreeMap<u64, BTreeMap<String, String>>,
    local_digests: BTreeMap<u64, String>,
}

impl LockstepCoordinator {
    pub fn new(
        mut required_peers: Vec<String>,
        definition_set_hash: String,
        input_availability_policy: &str,
        digest_interval_ticks: u64,
        desynchronization_policy: &str,
        predictor_id: Option<&str>,
    ) -> Result<Self, NetworkError> {
        required_peers.sort_by(|left, right| left.as_bytes().cmp(right.as_bytes()));
        if required_peers.is_empty()
            || required_peers.windows(2).any(|pair| pair[0] == pair[1])
            || !matches!(input_availability_policy, "WAIT" | "PREDICT")
            || digest_interval_ticks == 0
            || !matches!(
                desynchronization_policy,
                "pcam.desync.abort" | "pcam.desync.report"
            )
            || (input_availability_policy == "PREDICT"
                && predictor_id != Some("pcam.predict.no-input.v1"))
        {
            return Err(NetworkError::InvalidProfile);
        }
        Ok(Self {
            required_peers,
            definition_set_hash,
            input_availability_policy: input_availability_policy.to_owned(),
            digest_interval_ticks,
            desynchronization_policy: desynchronization_policy.to_owned(),
            next_tick: 0,
            aborted: false,
            packets: BTreeMap::new(),
            digest_ticks: BTreeSet::new(),
            digest_reports: BTreeMap::new(),
            local_digests: BTreeMap::new(),
        })
    }

    pub fn submit(
        &mut self,
        peer_id: &str,
        tick: u64,
        definition_set_hash: &str,
        inputs: &[Value],
        contacts: &[Value],
    ) -> Result<(), NetworkError> {
        self.require_active()?;
        if !self.required_peers.iter().any(|peer| peer == peer_id) {
            return Err(NetworkError::UnknownPeer);
        }
        if tick != self.next_tick {
            return Err(NetworkError::TickMismatch);
        }
        if definition_set_hash != self.definition_set_hash {
            return Err(NetworkError::DefinitionMismatch);
        }
        if self.packets.contains_key(peer_id) {
            return Err(NetworkError::DuplicatePacket);
        }
        if inputs
            .iter()
            .any(|input| input["assigned_tick"].as_u64() != Some(tick))
        {
            return Err(NetworkError::InputTickMismatch);
        }
        for input in inputs {
            canonicalize(input)?;
        }
        self.packets.insert(
            peer_id.to_owned(),
            PeerPacket {
                inputs: inputs.to_vec(),
                contacts: contacts.to_vec(),
            },
        );
        Ok(())
    }

    pub fn advance(&mut self) -> Result<LockstepTick, NetworkError> {
        self.require_active()?;
        let missing = self
            .required_peers
            .iter()
            .filter(|peer| !self.packets.contains_key(*peer))
            .cloned()
            .collect::<Vec<_>>();
        if !missing.is_empty() && self.input_availability_policy == "WAIT" {
            return Ok(LockstepTick {
                status: "WAITING".to_owned(),
                tick: self.next_tick,
                tick_document: None,
                missing_peers: missing,
                predicted_peers: Vec::new(),
                digest_due: false,
            });
        }
        let packets = self
            .required_peers
            .iter()
            .filter_map(|peer| self.packets.get(peer))
            .collect::<Vec<_>>();
        let host_hashes = packets
            .iter()
            .map(|packet| canonical_hash(&Value::Array(packet.contacts.clone())))
            .collect::<Result<BTreeSet<_>, _>>()?;
        if host_hashes.len() > 1 {
            return Err(NetworkError::HostMismatch);
        }
        let contacts = packets
            .first()
            .map(|packet| packet.contacts.clone())
            .unwrap_or_default();
        let mut inputs = packets
            .iter()
            .flat_map(|packet| packet.inputs.iter().cloned())
            .collect::<Vec<_>>();
        let identifiers = inputs
            .iter()
            .map(|input| input["input_id"].as_str().unwrap_or_default())
            .collect::<Vec<_>>();
        if identifiers.len() != identifiers.iter().collect::<BTreeSet<_>>().len() {
            return Err(NetworkError::DuplicateInput);
        }
        inputs.sort_by(|left, right| input_key(left).cmp(&input_key(right)));
        let tick = self.next_tick;
        let digest_due = (tick + 1) % self.digest_interval_ticks == 0;
        self.packets.clear();
        self.next_tick += 1;
        if digest_due {
            self.digest_ticks.insert(tick);
        }
        Ok(LockstepTick {
            status: "READY".to_owned(),
            tick,
            tick_document: Some(json!({"inputs": inputs, "contacts": contacts})),
            missing_peers: Vec::new(),
            predicted_peers: missing,
            digest_due,
        })
    }

    pub fn submit_digest(
        &mut self,
        peer_id: &str,
        tick: u64,
        reported_digest: &str,
        local_digest: &str,
    ) -> Result<DigestResolution, NetworkError> {
        self.require_active()?;
        if !self.required_peers.iter().any(|peer| peer == peer_id) {
            return Err(NetworkError::UnknownPeer);
        }
        if !self.digest_ticks.contains(&tick) {
            return Err(NetworkError::DigestNotDue);
        }
        match self.local_digests.get(&tick) {
            Some(known) if known != local_digest => return Err(NetworkError::LocalDigestChanged),
            None => {
                self.local_digests.insert(tick, local_digest.to_owned());
            }
            _ => {}
        }
        let reports = self.digest_reports.entry(tick).or_default();
        if reports.contains_key(peer_id) {
            return Err(NetworkError::DuplicateDigest);
        }
        reports.insert(peer_id.to_owned(), reported_digest.to_owned());
        if reports.len() < self.required_peers.len() {
            return Ok(DigestResolution {
                status: "WAITING".to_owned(),
                tick,
                mismatched_peers: Vec::new(),
            });
        }
        let mismatched = self
            .required_peers
            .iter()
            .filter(|peer| reports[*peer] != local_digest)
            .cloned()
            .collect::<Vec<_>>();
        let status = if mismatched.is_empty() {
            "MATCH"
        } else if self.desynchronization_policy == "pcam.desync.abort" {
            self.aborted = true;
            "ABORTED"
        } else {
            "REPORTED"
        };
        Ok(DigestResolution {
            status: status.to_owned(),
            tick,
            mismatched_peers: mismatched,
        })
    }

    fn require_active(&self) -> Result<(), NetworkError> {
        if self.aborted {
            Err(NetworkError::Aborted)
        } else {
            Ok(())
        }
    }
}

#[derive(Debug, Clone)]
pub struct ServerAuthoritativeCorrectionPlanner {
    correction_policy: String,
    max_latency_compensation_ticks: u64,
}

impl ServerAuthoritativeCorrectionPlanner {
    pub fn new(
        correction_policy: &str,
        max_latency_compensation_ticks: u64,
    ) -> Result<Self, NetworkError> {
        if !matches!(
            correction_policy,
            "pcam.correction.resimulate.v1" | "pcam.correction.replace-discard.v1"
        ) {
            return Err(NetworkError::InvalidProfile);
        }
        Ok(Self {
            correction_policy: correction_policy.to_owned(),
            max_latency_compensation_ticks,
        })
    }

    pub fn plan(
        &self,
        current_tick: u64,
        authoritative_tick: u64,
        complete_state_supplied: bool,
    ) -> Result<ServerCorrectionPlan, NetworkError> {
        if authoritative_tick > current_tick {
            return Err(NetworkError::InvalidCorrection);
        }
        let discard = current_tick - authoritative_tick;
        if discard > self.max_latency_compensation_ticks {
            return Err(NetworkError::CorrectionWindowExceeded);
        }
        let operation = if self.correction_policy == "pcam.correction.replace-discard.v1" {
            if !complete_state_supplied {
                return Err(NetworkError::CompleteStateRequired);
            }
            "REPLACE_AND_DISCARD"
        } else {
            "RESTORE_AND_RESIMULATE"
        };
        Ok(ServerCorrectionPlan {
            operation: operation.to_owned(),
            authoritative_tick,
            discard_prediction_ticks: discard,
        })
    }
}

fn input_key(value: &Value) -> (u64, u64, Vec<u8>, Vec<u8>, Vec<u8>) {
    (
        value["source_entity_id"].as_u64().unwrap_or_default(),
        value["sequence"].as_u64().unwrap_or_default(),
        value["command_id"]
            .as_str()
            .unwrap_or_default()
            .as_bytes()
            .to_vec(),
        value["input_id"]
            .as_str()
            .unwrap_or_default()
            .as_bytes()
            .to_vec(),
        canonicalize(value).unwrap_or_default(),
    )
}
